src/macos_native.rs

```diff
@@ -0,0 +1,892 @@
+    _result: IOReturn,
+    _sender: *mut c_void,
+    _report_type: kIOHIDReportType,
+    _report_id: u32,
+    report: *mut u8,
+    report_length: CFIndex,
+) {
+    let shared_state = unsafe { &*(context as *const SharedState) };
+
+    let data = unsafe { std::slice::from_raw_parts(report, report_length as usize) };
+
+    let mut input_reports = shared_state.input_reports.lock().unwrap();
+
+    // Ensure there are never more than 30 reports in the queue
+    // Copied from hidapi
+    if input_reports.len() == 30 {
+        input_reports.pop_front();
+    }
+
+    input_reports.push_back(data.to_vec());
+
+    shared_state.condition.notify_one();
+}
+
+fn read_thread_fun(barrier: Arc<Barrier>, shared_state: Arc<SharedState>) {
+    // This must live as long as the callback is registered
+    let mut input_report_buffer = vec![0u8; shared_state.max_input_report_len];
+
+    // This must live as long as the callback is registered
+    let ctx_ptr = Arc::as_ptr(&shared_state) as *const c_void as *mut c_void;
+
+    let input_report_context = shared_state.clone();
+
+    let run_loop_mode = CFString::new(&shared_state.run_loop_mode);
+
+    let mut _input_report_callback = None;
+
+    {
+        let device = shared_state.device.lock().unwrap();
+
+        _input_report_callback = Some(device.register_input_report_callback(
+            &mut input_report_buffer,
+            Some(hid_report_callback),
+            input_report_context,
+        ));
+
+        unsafe { device.register_removal_callback(Some(hid_removal_callback), ctx_ptr) }
+
+        device.schedule_with_run_loop(&CFRunLoop::get_current(), &run_loop_mode);
+
+        let mut ctx = CFRunLoopSourceContext {
+            version: 0,
+            info: ctx_ptr as *const c_void as *mut _,
+            retain: None,
+            release: None,
+            copyDescription: None,
+            equal: None,
+            hash: None,
+            schedule: None,
+            cancel: None,
+            perform: perform_signal_callback,
+        };
+
+        let source = unsafe {
+            CFRunLoopSourceCreate(kCFAllocatorDefault, 0 /* order */, &mut ctx)
+        };
+
+        let source = unsafe { CFRunLoopSource::wrap_under_create_rule(source) };
+
+        let current_run_loop = CFRunLoop::get_current();
+
+        current_run_loop.add_source(&source, run_loop_mode.as_concrete_TypeRef());
+
+        let mut shared_source = shared_state.source.lock().unwrap();
+        *shared_source = Some(LoopSource(source));
+    }
+
+    {
+        let mut run_loop = shared_state.run_loop.lock().unwrap();
+
+        *run_loop = Some(WrappedCFRunLoop(CFRunLoop::get_current()));
+    }
+
+    barrier.wait();
+
+    while (!shared_state.shutdown_thread.load(Ordering::Relaxed))
+        && (!shared_state.disconnected.load(Ordering::Relaxed))
+    {
+        // TODO: Verify timeout value
+        let code = unsafe { CFRunLoopRunInMode(run_loop_mode.as_concrete_TypeRef(), 1000.0, 0) };
```

Comment:
I'm struggling to understand the design here, but why not just `CFRunLoopRun()`? It does basically the same thing as your loop: https://github.com/apple-oss-distributions/CF/blob/dc54c6bb1c1e5e0b9486c1d26dd5bef110b20bf3/CFRunLoop.c#L2676-L2682

On disconnect/`Drop`, you can issue a `CFRunLoopStop()` from the main thread – `HidDevice::drop` already has a reference to the device's `CFRunLoop`.

I don't think it matters in this loop whether the device was closed by the application (terminating the run loop) or there was an explicit disconnect (terminating the run loop); a disconnected device is a disconnected device. :)

`mpsc::channel` also saves you here too, because `Drop`ing the `Sender` will unblock any pending `Receiver::recv` calls on another thread, and make future calls fail.

If you've only associated a single `IOHIDDevice` with the `CFRunLoop`, the run loop will stop on disconnect _anyway_. `hid_removal_callback` appears to only set a flag that it was disconnected, but this could _also_ be detected by setting your own (not-mutex-guarded) flag on `close()`.